File: world/map.go
// Package world 負責地圖格點、地形、阻擋與移動碰撞。本檔為地圖與地形，對齊 spec_移動與地圖。
// 第一版可做清單 1.2.1：大地圖格點資料（Grid）、地形字（Terrain）、阻擋表（Blocking）；經 1.2.3 區塊載入與 ChunkView 使用。
package world

// Terrain 代表單格地形類型，大地圖只存此一字；對齊 map_terrain_world、地形字對照表。
type Terrain string

const (
	TerrainWall       Terrain = "牆"
	TerrainDoor       Terrain = "門"
	TerrainDoorClosed Terrain = "關"
	TerrainRoad       Terrain = "道"
	TerrainAlley      Terrain = "巷" // 房屋間距小成巷，與道／地同色系（城市底色）
	TerrainGrass    Terrain = "草"
	TerrainTree     Terrain = "木"
	TerrainMountain Terrain = "山"
	TerrainStone    Terrain = "石"
	TerrainSwamp    Terrain = "沼"
	TerrainRiver    Terrain = "川"
	TerrainWater    Terrain = "水"
	TerrainWasteland Terrain = "荒"
	TerrainLava     Terrain = "火"
	TerrainIce      Terrain = "冰"
	TerrainFarm     Terrain = "田"
	TerrainValley   Terrain = "谷"
	TerrainFog      Terrain = "霧"
	TerrainFloor    Terrain = "地" // 房屋內部地板，單一字，顏色與當前城市底色相同
)

// Blocking 回傳該地形是否阻擋移動。牆、關、川、水阻擋；門（開門）、道、草等可通行。
func (t Terrain) Blocking() bool {
	return t == TerrainWall || t == TerrainDoorClosed || t == TerrainRiver || t == TerrainWater
}

// Grid 代表大地圖格點資料，每格一個地形；人物尺寸＝一格。
type Grid struct {
	Width   int
	Height  int
	Cells   []Terrain
}

// At 回傳 (x,y) 格的地形；越界視為牆（阻擋）。
func (g *Grid) At(x, y int) Terrain {
	if x < 0 || x >= g.Width || y < 0 || y >= g.Height {
		return TerrainWall
	}
	return g.Cells[y*g.Width+x]
}

// NewGrid 建立寬高為 w,h 的空地圖，預設為草。
func NewGrid(w, h int) *Grid {
	cells := make([]Terrain, w*h)
	for i := range cells {
		cells[i] = TerrainGrass
	}
	return &Grid{Width: w, Height: h, Cells: cells}
}

File: src/world/map.rs
// 地圖格點與地形，對齊既有 world/map。

/// 牆
pub const TERRAIN_WALL: &str = "牆";
/// 門
pub const TERRAIN_DOOR: &str = "門";
/// 關
pub const TERRAIN_DOOR_CLOSED: &str = "關";
/// 道
pub const TERRAIN_ROAD: &str = "道";
/// 巷
pub const TERRAIN_ALLEY: &str = "巷";
/// 草
pub const TERRAIN_GRASS: &str = "草";
/// 木
pub const TERRAIN_TREE: &str = "木";
/// 山
pub const TERRAIN_MOUNTAIN: &str = "山";
/// 石
pub const TERRAIN_STONE: &str = "石";
/// 沼
pub const TERRAIN_SWAMP: &str = "沼";
/// 川
pub const TERRAIN_RIVER: &str = "川";
/// 水
pub const TERRAIN_WATER: &str = "水";
/// 荒
pub const TERRAIN_WASTELAND: &str = "荒";
/// 火
pub const TERRAIN_LAVA: &str = "火";
/// 冰
pub const TERRAIN_ICE: &str = "冰";
/// 田
pub const TERRAIN_FARM: &str = "田";
/// 谷
pub const TERRAIN_VALLEY: &str = "谷";
/// 霧
pub const TERRAIN_FOG: &str = "霧";
/// 地（屋內地板）
pub const TERRAIN_FLOOR: &str = "地";

/// 該地形字是否阻擋移動（牆、關、川、水）。
pub fn terrain_blocking(t: &str) -> bool {
    t == TERRAIN_WALL || t == TERRAIN_DOOR_CLOSED || t == TERRAIN_RIVER || t == TERRAIN_WATER
}

/// 大地圖格點：每格一個地形一字（UTF-8 字串）。
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Grid {
    pub width: usize,
    pub height: usize,
    pub cells: Vec<String>,
}

impl Grid {
    /// 建立 w×h 地圖，預設為草。
    pub fn new(w: usize, h: usize) -> Self {
        let mut cells = Vec::with_capacity(w * h);
        cells.resize(w * h, TERRAIN_GRASS.to_string());
        Self {
            width: w,
            height: h,
            cells,
        }
    }

    /// 越界視為牆。
    pub fn at(&self, x: i32, y: i32) -> &str {
        if x < 0 || y < 0 {
            return TERRAIN_WALL;
        }
        let (x, y) = (x as usize, y as usize);
        if x >= self.width || y >= self.height {
            return TERRAIN_WALL;
        }
        &self.cells[y * self.width + x]
    }

    pub fn set_at(&mut self, x: usize, y: usize, terrain: String) {
        if x < self.width && y < self.height {
            self.cells[y * self.width + x] = terrain;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn blocking_matches_legacy() {
        assert!(terrain_blocking(TERRAIN_WALL));
        assert!(terrain_blocking(TERRAIN_DOOR_CLOSED));
        assert!(terrain_blocking(TERRAIN_RIVER));
        assert!(terrain_blocking(TERRAIN_WATER));
        assert!(!terrain_blocking(TERRAIN_DOOR));
        assert!(!terrain_blocking(TERRAIN_ROAD));
    }

    #[test]
    fn at_out_of_bounds_is_wall() {
        let g = Grid::new(3, 3);
        assert_eq!(g.at(-1, 0), TERRAIN_WALL);
        assert_eq!(g.at(3, 0), TERRAIN_WALL);
    }
}
